File: gsim/core_gateway/schema.py
"""
IRS message class  ->  JSON schema the UI can render a form from.

This is pure introspection: it reads `Message._fields_` (built by
`IRS.core.MessageMeta` at class-definition time) and never mutates anything in
`core`. Every field kind the IRS metaclass can produce is covered:

    Field       -> {"kind": "scalar"}    an int/float, dtype named (UInt16, ...)
    EnumField   -> {"kind": "enum"}      renders as a <select>
    Structure   -> {"kind": "struct"}    nested fields, rendered recursively
    BitField    -> {"kind": "bitfield"}  packed bits; each bit-range is an input
    ArrayField  -> {"kind": "array"}     see `length` below

`ArrayField.length` is the load-bearing distinction for the UI:

    None    -> "dynamic"  consumes the rest of the buffer; grows by user action
    str     -> "counted"  its length lives in a SIBLING field of that name
    int     -> "fixed"    exactly N items, UI must not offer add/remove

The "counted" case needs care on the way out: `ArrayField.to_bytes` iterates the
list and never writes the count itself, while `from_bytes` reads exactly
`getattr(instance, length_field)` items. A count that disagrees with the list is
therefore silent wire corruption, not an error -- `payloads.py` keeps the two in
sync, which is why `count_field` is carried in the schema.
"""
from __future__ import annotations

from enum import IntEnum
from typing import Any

from . import bootstrap  # noqa: F401  -- must precede every `core` import

import core.IRS.constants as _constants
from core.IRS.bitfields import BitField
from core.IRS.core import ArrayField, Structure
from core.IRS.fields import BaseField, EnumField, Field

#: struct format char -> the name the IRS constants module gives it. Built from
#: `IRS.constants` rather than hardcoded, so a new type there shows up here for
#: free. `B` is both `UInt8` and `Byte`; first-wins keeps it at `UInt8`.
_DTYPE_BY_CHAR: dict[str, str] = {}
for _name, _fmt in vars(_constants).items():
    if not _name.startswith("_") and isinstance(_fmt, str) and len(_fmt) == 1:
        _DTYPE_BY_CHAR.setdefault(_fmt, _name)

_FLOAT_CHARS = frozenset("fd")
_BYTE_WIDTH = {"b": 1, "B": 1, "h": 2, "H": 2, "i": 4, "I": 4, "q": 8, "Q": 8, "f": 4, "d": 8}


def _format_char(packer) -> str:
    """The bare type char out of a `struct.Struct` format, minus byte-order."""
    return packer.format.lstrip("<>=!@")


#: `struct`'s byte-order prefixes. `=` and `@` mean native, which on every
#: platform this runs on is little-endian; an absent prefix means `@`.
_BIG_ENDIAN_PREFIXES = frozenset(">!")


def _byte_order(packer) -> str:
    """Which way round this field's bytes go, read off its own packer.

    Carried into the schema so the UI's byte view can render the wire order it
    will actually have instead of assuming one. Reading it here rather than
    hardcoding "little" costs a character comparison and means a layout that
    ever declares network order shows up correctly rather than reversed.
    """
    prefix = packer.format[:1]
    return "big" if prefix in _BIG_ENDIAN_PREFIXES else "little"


def _numeric_meta(char: str, endian: str = "little") -> dict[str, Any]:
    """dtype name, wire width, byte order, and the inclusive range -- so the UI
    can bound its inputs, reject a bad value before it ever reaches
    `struct.pack`, total up the message length it is about to send, and show
    the bytes that message will occupy."""
    width = _BYTE_WIDTH.get(char, 1)
    meta: dict[str, Any] = {
        "dtype": _DTYPE_BY_CHAR.get(char, char),
        "byte_size": width,
        "endian": endian,
    }
    if char in _FLOAT_CHARS:
        meta["numeric"] = "float"
        return meta
    bits = width * 8
    signed = char.islower()
    meta["numeric"] = "int"
    meta["min"] = -(1 << (bits - 1)) if signed else 0
    meta["max"] = (1 << (bits - 1)) - 1 if signed else (1 << bits) - 1
    return meta


def _enum_options(enum_class: type[IntEnum]) -> list[dict[str, Any]]:
    return [{"name": member.name, "value": member.value} for member in enum_class]


def _bitfield_schema(name: str, bitfield: BitField) -> dict[str, Any]:
    """A BitField packs several small ints into one integer. `_fields_` is
    (name, shift, bits) and `__enum_map__` names the ones that are enums."""
    cls = type(bitfield)
    enum_map = getattr(cls, "__enum_map__", {})
    bits: list[dict[str, Any]] = []
    for bit_name, shift, width in cls._fields_:
        entry: dict[str, Any] = {
            "name": bit_name,
            "shift": shift,
            "bits": width,
            "min": 0,
            "max": (1 << width) - 1,
        }
        enum_class = enum_map.get(bit_name)
        if enum_class is not None:
            entry["kind"] = "enum"
            entry["enum"] = enum_class.__name__
            entry["options"] = _enum_options(enum_class)
        else:
            entry["kind"] = "scalar"
            entry["numeric"] = "int"
        bits.append(entry)
    # The whole bitfield occupies one packed integer, whose width comes from
    # its @baseType -- not from the sum of its bit widths, which may be less.
    packer = getattr(cls, "_packer_", None)
    return {
        "name": name,
        "kind": "bitfield",
        "struct": cls.__name__,
        "byte_size": packer.size if packer is not None else 1,
        "endian": _byte_order(packer) if packer is not None else "little",
        "bits": bits,
    }


def describe_field(field: BaseField, name: str | None = None) -> dict[str, Any]:
    """One `_fields_` entry -> one schema node. Recursive for struct/array."""
    field_name = name if name is not None else field._name

    # Order matters: BitField and Structure are both BaseField subclasses and
    # ArrayField is checked before them only because it is never either.
    if isinstance(field, ArrayField):
        item = describe_field(field.baseType, name="item")
        node: dict[str, Any] = {"name": field_name, "kind": "array", "item": item}
        if field.length is None:
            node["length"] = "dynamic"
        elif isinstance(field.length, str):
            node["length"] = "counted"
            node["count_field"] = field.length
        else:
            node["length"] = "fixed"
            node["size"] = int(field.length)
        # What the "Add <X>" button should say.
        node["item_label"] = item.get("struct") or item.get("dtype") or "item"
        return node

    if isinstance(field, BitField):
        return _bitfield_schema(field_name, field)

    if isinstance(field, Structure):
        return {
            "name": field_name,
            "kind": "struct",
            "struct": type(field).__name__,
            "fields": [describe_field(sub) for sub in type(field)._fields_],
        }

    if isinstance(field, EnumField):
        return {
            "name": field_name,
            "kind": "enum",
            "enum": field.enum_class.__name__,
            "options": _enum_options(field.enum_class),
            **_numeric_meta(_format_char(field.packer), _byte_order(field.packer)),
        }

    if isinstance(field, Field):
        return {
            "name": field_name,
            "kind": "scalar",
            **_numeric_meta(_format_char(field.packer), _byte_order(field.packer)),
        }

    # An IRS field kind this module has not been taught yet. Surface it as
    # read-only rather than guessing at a widget for it.
    return {"name": field_name, "kind": "unsupported", "python_type": type(field).__name__}


def describe_message(message_class: type, unit_code: int, op_code: int) -> dict[str, Any]:
    """Full form schema for one registered (unitCode, opCode) message."""
    return {
        "unit_code": unit_code,
        "op_code": op_code,
        "name": message_class.__name__,
        "doc": (message_class.__doc__ or "").strip() or None,
        "fields": [describe_field(field) for field in message_class._fields_],
    }


#: Field kinds a filter RULE can compare against. A rule needs a single value to
#: put on one side of an operator, which a struct, a bitfield-as-a-whole, or an
#: array does not have.
_COMPARABLE = frozenset({"scalar", "enum"})


def filter_targets(schema: dict[str, Any]) -> list[dict[str, Any]]:
    """Every field of one message a received-filter may address, flattened.

    Two different questions, answered per node, because the two halves of a
    filter can address different things:

    `rule_ok`   -- may a keep/drop rule compare this? Only a single decoded
                   value: a scalar, an enum, or one bit of a bitfield.
    `change_ok` -- may the change trigger watch this? Anything reachable,
                   including a whole struct, a whole bitfield, or an array --
                   the trigger compares by value, and `==` on a list or a dict
                   is exactly the right question there.

    **The walk never descends into an array's item.** A dotted path cannot name
    one of 35 elements, and silently addressing the first would be a lie the
    user could not see -- so an array is offered as a change subject (compare
    the whole list) and nothing inside it is offered at all. This is the one
    place that limitation is decided; `filters.py` only ever sees paths.

    Paths use the same dotted grammar as the byte view's `leafOffsets`
    (`Position.Latitude`), so a field is spelt one way across the whole UI.
    """
    targets: list[dict[str, Any]] = []

    def emit(path: list[str], node: dict[str, Any], rule_ok: bool) -> None:
        target: dict[str, Any] = {
            "path": ".".join(path),
            "name": path[-1],
            "kind": node.get("kind", "scalar"),
            "rule_ok": rule_ok,
        }
        for key in ("dtype", "enum", "options", "min", "max", "numeric", "struct",
                    "item_label", "length"):
            if key in node:
                target[key] = node[key]
        targets.append(target)

    def walk(node: dict[str, Any], path: list[str]) -> None:
        kind = node.get("kind")
        if kind == "struct":
            emit(path, node, rule_ok=False)
            for child in node.get("fields", []):
                walk(child, [*path, child["name"]])
            return
        if kind == "bitfield":
            emit(path, node, rule_ok=False)
            for bit in node.get("bits", []):
                # A bit IS comparable -- `BitField.to_dict` emits one key per
                # bit, with enum bits as member names, so `Area.fr` resolves.
                emit([*path, bit["name"]], bit, rule_ok=True)
            return
        if kind == "array":
            emit(path, node, rule_ok=False)     # watchable, never rule-addressable
            return
        emit(path, node, rule_ok=kind in _COMPARABLE)

    for field_node in schema.get("fields", []):
        walk(field_node, [field_node["name"]])
    return targets

File: gsim/web/src/lib/schema.js
/**
 * Client-side mirror of gsim/core_gateway/payloads.py.
 *
 * The server re-normalises every payload before encoding, so this is not a
 * trust boundary -- it exists so the form starts out fully populated (every
 * input controlled from first render) and so the user can see exactly what
 * will be sent. The two must agree on defaults; if they ever drift, the
 * server wins.
 */

/** The zero value for one schema node. */
export function defaultFor(node) {
  switch (node.kind) {
    case 'scalar':
      return node.numeric === 'float' ? 0.0 : 0;
    case 'enum': {
      const options = node.options ?? [];
      // Prefer the member whose value is 0 (the conventional UNKNOWN/unset
      // member); EnumField.from_bytes rejects any value with no member.
      const zero = options.find((option) => option.value === 0);
      return zero ? zero.value : options[0]?.value ?? 0;
    }
    case 'struct':
      return Object.fromEntries(node.fields.map((f) => [f.name, defaultFor(f)]));
    case 'bitfield':
      return Object.fromEntries(node.bits.map((b) => [b.name, defaultFor(b)]));
    case 'array':
      return node.length === 'fixed'
        ? Array.from({ length: node.size ?? 0 }, () => defaultFor(node.item))
        : [];
    default:
      return 0;
  }
}

/** Initial form state for a whole message schema. */
export function defaultPayload(schema) {
  return Object.fromEntries(schema.fields.map((f) => [f.name, defaultFor(f)]));
}

/**
 * Names of scalar fields that are really an array's length counter.
 *
 * IRS does NOT write these itself -- `ArrayField.to_bytes` just iterates the
 * list, while `from_bytes` reads exactly `getattr(instance, count_field)`
 * items. So the count is derived from the list, and the UI renders it
 * read-only rather than letting the user desync the wire format by hand.
 */
export function derivedCounters(fields) {
  return new Set(
    fields
      .filter((f) => f.kind === 'array' && f.length === 'counted' && f.count_field)
      .map((f) => f.count_field),
  );
}

/** The array a given counter field belongs to, for the read-only display. */
export function counterSource(fields, counterName) {
  return fields.find(
    (f) => f.kind === 'array' && f.length === 'counted' && f.count_field === counterName,
  );
}

//: The framing header's DataLength is uint16 (`struct.Struct("<BHH")` in
//: core/connections/framing.py), so no payload can exceed this many bytes --
//: the ceiling on a dynamic array that has no counter to bound it.
const MAX_PAYLOAD_BYTES = 0xffff;

/**
 * How many items a growable array may hold, or `null` when nothing bounds it.
 *
 * Two different ceilings, because the two array kinds are limited by different
 * things:
 *
 *   counted  its length travels in a SIBLING scalar, so the array can never be
 *            longer than that field can count -- a UInt8 counter caps it at
 *            255 no matter how much frame is left. Going past that does not
 *            error, it silently truncates on the receiving side, which is the
 *            worst possible failure mode.
 *   dynamic  no counter exists (it fills the rest of the frame), so the only
 *            real bound is what the uint16 DataLength can describe.
 *
 * `fields` is the array's SIBLING list -- the counter is a sibling, not a
 * child, so this cannot be derived from the array node alone.
 */
export function maxArrayItems(node, fields) {
  if (node.kind !== 'array' || node.length === 'fixed') return null;

  if (node.length === 'counted') {
    const counter = (fields ?? []).find((f) => f.name === node.count_field);
    if (typeof counter?.max === 'number') return counter.max;
  }

  const itemBytes = sizeOf(node.item, defaultFor(node.item));
  if (itemBytes > 0) return Math.floor(MAX_PAYLOAD_BYTES / itemBytes);
  return null;
}

/**
 * Wire size in bytes of one field given its current value.
 *
 * Computed from the payload rather than the schema alone because a message's
 * length is not fixed: dynamic and counted arrays contribute
 * `items x item size`, so the total only exists once you know how many items
 * there are. Scalars/enums carry `byte_size` from the struct format char, and
 * a bitfield carries the width of its packed integer (its @baseType), which is
 * not the sum of its bit widths.
 */
function sizeOf(node, value) {
  switch (node.kind) {
    case 'scalar':
    case 'enum':
      return node.byte_size ?? 0;
    case 'bitfield':
      return node.byte_size ?? 0;
    case 'struct':
      return node.fields.reduce((sum, f) => sum + sizeOf(f, value?.[f.name]), 0);
    case 'array': {
      const items = Array.isArray(value) ? value : [];
      return items.reduce((sum, item) => sum + sizeOf(node.item, item), 0);
    }
    default:
      return 0;
  }
}

/** Total payload size for a message, excluding the framing header. */
export function payloadSize(fields, payload) {
  return fields.reduce((sum, field) => sum + sizeOf(field, payload?.[field.name]), 0);
}

/**
 * How many characters wide a field's input needs to be.
 *
 * Every scalar used to render full-width, one per row, whatever it could hold:
 * a `UInt8` bounded at 255 got the same ~900px box as a `Double`, so a
 * seven-field message scrolled while holding seven zeros. The schema already
 * says exactly how much room each value needs -- `max` gives the digit count
 * and `numeric` says whether a decimal point and exponent are possible -- so
 * the control can simply be that wide.
 *
 * Returned in `ch`, and deliberately generous: a value being typed can be
 * momentarily longer than the field's range allows (a leading minus, a pasted
 * number about to be trimmed), and a control that clips mid-edit is worse than
 * one with a little slack.
 */
export function controlChars(node) {
  if (node.kind === 'enum') return 16;
  if (node.numeric === 'float') return 16;
  if (typeof node.max === 'number') {
    const digits = String(node.max).length;
    const sign = typeof node.min === 'number' && node.min < 0 ? 1 : 0;
    return Math.min(16, Math.max(6, digits + sign + 2));
  }
  if (typeof node.bits === 'number') return Math.max(6, String((1 << node.bits) - 1).length + 2);
  return 10;
}

/**
 * Byte offsets for every leaf in a message, flattened in wire order.
 *
 * The byte ruler needs to map a field to a byte range and back, and that map is
 * pure arithmetic over the schema plus the current payload -- `sizeOf` above
 * already computes each node's width, so walking the tree in `_fields_` order
 * and accumulating gives the offset of every leaf. It depends on the payload
 * because a message's length is not fixed: a counted or dynamic array
 * contributes `items x item size`, so nothing after one has a knowable offset
 * until you know how many items there are.
 *
 * `path` is the dotted name shown to the user (`Position.Latitude`,
 * `Emitters[0].PowerDbm`); `key` is the same path as an array, which is what
 * the renderer compares against to know whether a given field is the one lit.
 */
export function leafOffsets(fields, payload) {
  const out = [];
  let offset = 0;

  const walk = (node, value, path) => {
    switch (node.kind) {
      case 'struct':
        for (const child of node.fields) {
          walk(child, value?.[child.name], [...path, child.name]);
        }
        return;
      case 'array': {
        const items = Array.isArray(value) ? value : [];
        items.forEach((item, index) => {
          // The index belongs to the LAST segment, not a segment of its own --
          // `Emitters[0]`, so the label reads the way the field is written.
          const head = path.slice(0, -1);
          const tail = `${path[path.length - 1]}[${index}]`;
          walk(node.item, item, [...head, tail]);
        });
        return;
      }
      case 'bitfield':
      case 'scalar':
      case 'enum': {
        const size = sizeOf(node, value);
        out.push({
          path: path.join('.'),
          key: path.join('.'),
          kind: node.kind,
          node,
          value,
          offset,
          size,
        });
        offset += size;
        return;
      }
      default:
        return;
    }
  };

  for (const field of fields) walk(field, payload?.[field.name], [field.name]);
  return out;
}
